Activate Early backoff functionality (#253)
gardener/autoscaler: cluster-autoscaler/FAQ.md

  * [How can I update CA dependencies (particularly k8s.io/kubernetes)?](#how-can-i-update-ca-dependencies-particularly-k8siokubernetes)

* [In the context of Gardener](#in-the-context-of-gardener)
  * [For User](#for-user)

>     * [When does autoscaler backs off early from a node group?](#when-does-autoscaler-backs-off-early-from-a-node-group)

Reviewer: ```suggestion
    * [When does autoscaler back off early from a node group?](#when-does-autoscaler-backs-off-early-from-a-node-group)
```

# In the context of Gardener:

## For User
### When does autoscaler backs off early from a node group?

Autoscaler backs off from a node group if the scale-up requested doesn't succeed. Autoscaler decides to backoff based on:
- Timeout
  - if the node doesn't join in `max-node-provision-time`
- `ResourceExhausted` error 
  - if the node doesn't join due to error from cloud provider side, and the error is classified as `ResourceExhausted`
- Scale up operation fails for a node group

> Early backoff , as the name suggests doesn't wait till `timeout` , but backsoff as soon as certain condition is satisfied. This helps in trying other node groups quickly.

Reviewer: ```suggestion
As the name suggests, early back-off doesn't wait till `timeout` but backs off when a certain condition is satisfied. This helps in trying other node groups quickly.
```

Currently early-backoff is enabled only for `ResourceExhausted` errors. Errors classified as `ResourceExhausted` are(and not limited to):
- `out of quota` errors where customer quota is exhausted, and the quota is configurable per zone (not per region). Generally quotas for VMs, cpus, gpus and disks are configurable per zone, but please confirm the same for your cloud provider
- `out of stock` errors where cloud-provider doesn't have enough resources in the particular zone, but the resource is available in other zones
- `not-supported` errors where the instance type or disk type is not supported in the particular zone. 

Errors not classified as `ResourceExhausted` are:(and not limited to):
- `invalid credentials`
- `rate limiting`
- `policy constraints defined by customer`
- `service-unavailable` on cloud-provider side

> Backoff after `timeout` will happen for these errors.

Reviewer: ```suggestion
Backoff after `timeout` will happen for errors other than `ResourceExhausted`.
```

- `out of quota` errors where customer quota is exhausted, and the quota is configurable per zone (not per region). Generally quotas for VMs, cpus, gpus and disks are configurable per zone, but please confirm the same for your cloud provider
- `out of stock` errors where cloud-provider doesn't have enough resources in the particular zone, but the resource is available in other zones
- `not-supported` errors where the instance type or disk type is not supported in the particular zone. 

Errors not classified as `ResourceExhausted` are:(and not limited to):
- `invalid credentials`
- `rate limiting`
- `policy constraints defined by customer`
- `service-unavailable` on cloud-provider side

Backoff after `timeout` will happen for errors other than `ResourceExhausted`.

*NOTE:* The identifier for the error might differ for each cloud-provider. The above listed errors are general names used.

**--Caveat during rolling update--**

Case:

- If node-grp `ng-A` is in rolling update, AND
- If the scale-up happens for `ng-A` due to an unschedulable pod `podA`, or a set of pods, AND
- if the node(say `node1`) couldn't join due to `ResourceExhausted`

> then autoscaler will early backoff and try to remove the node, but the node removal won't succeed as currently CA is not allowed to perform any scale-down/delete node operation for a rolling update node-grp.

Reviewer 2: First of all the terminology is incorrect. `Scale-up` means having a higher sized VM and not launching another VM that is considered as `Scale-Out`. Similarly `Scale-down` is downsizing the VM size and not the number of VMs and should instead be called `Scale-In`.

**There is a possible way to mitigate this:**
During rolling update, if a `Scale-Out` is allowed then all such machines should be annotated/labeled. This enables to allow `Scale-In` during rolling update for machines which failed to be provisioned and were launched due to `Scale-Out` during a rolling update. So if a rolling update is in progress and a request to downsize a node group is made by CA, then mcm-provider can check it is the machine that was launched as a result of `Scale-Out` and is now cannot be launched due to `ResourceExhausted` then it allows it, else the request is rejected. Can this work?

Reviewer 3: Term is sadly the autoscaler terminology since it scales up and scales-down node groups. We need to generate a `correlationId` for each `scale-up` request triggered by the autoscaler. The annotations should include this `correlationId` along with the time the scale-up request was triggered.

Author: as discussed, the solution proposed would require changes in MCM and need more analysis.. that shouldn't be a blocker for this PR.

In the above scenario, CA won't try to scale-up any other node-grp for `podA` as it still calculates `node1` to be a possible candidate to join(`ResourceExhausted` errors are recoverable errors). 

> Scale up would still work for any other new pods which can't fit on upcoming `node1`

Reviewer: ```suggestion
Scale-up would still work for any new pods that can't fit on upcoming `node1` but can fit on some other node group.

> If you are sure that the capacity won't recover soon, then kindly re-create `podA`. This will allow CA to see it as a new pod and allow scale-up.

Reviewer: This is not true; CA will still consider the node `node1` for the pod `podA`, and no scale-up will happen.
```suggestion

Author: I am considering the case that `ng-A` will be in backoff at this time, so scale-up won't happen in that again... ok I'll update that statement to mention this.

Author: done in [0309ec8](https://github.com/gardener/autoscaler/pull/253/commits/0309ec8e6f4a6da2a4ae21f0158f77c86f95c08d)
